pkg/endpoints/router.go

```diff
@@ -98,6 +113,72 @@ func (r Resource) RegisterReadinessProbes(container *restful.Container) {
 	container.Add(wsv4)
 }
 
+// Back-end extension: Requests to the URL are passed through to the Port of the Name service (extension)
+// "label: tekton-dashboard-extension=true" in the service defines the extension
+// "annotation: tekton-dashboard-endpoints=<URL>" specifies the path for the extension
+type Extension struct {
+	Name           string  `json:"name"`
+	URL            string  `json:"url"`
+	Port           string  `json:"port"`
+	DisplayName    string  `json:"displayname"`
+	BundleLocation string  `json:"bundlelocation"`
+}
+var	extensions     []Extension
+
+// RegisterExtension - this discovers the extensions and registers them as the REST API extension 
+func (r Resource) RegisterExtension(container *restful.Container, namespace string) {
+	logging.Log.Info("Adding API for extensions")
+	svcs, err := r.K8sClient.CoreV1().Services(namespace).List(metav1.ListOptions{LabelSelector: extensionLabel})
+	if err != nil {
+		logging.Log.Errorf("no extension found: %s", err)
+		return
+	}
+	ws := new(restful.WebService)
+	ws.
+		Path("/").
+		Consumes(restful.MIME_JSON).
+		Produces(restful.MIME_JSON)
+
+	for _, svc := range svcs.Items {
+		url, ok := svc.ObjectMeta.Annotations[urlKey]
+		if ok {
+			logging.Log.Debugf("extension URL: %s", url)
+			ext := Extension { Name: svc.ObjectMeta.Name, URL: url, Port: getPort(svc),
+ 				DisplayName: svc.ObjectMeta.Annotations[displayNameKey],
+				BundleLocation: svc.ObjectMeta.Annotations[bundleLocationKey]}
+			// extension handler is registered at the url
+			ws.Route(ws.GET(url).To(ext.HandleExtension))
+			ws.Route(ws.POST(url).To(ext.HandleExtension))
+			ws.Route(ws.PUT(url).To(ext.HandleExtension))
+			ws.Route(ws.DELETE(url).To(ext.HandleExtension))
+			extensions = append(extensions, ext)
+		}
+	}
```

Comment:
DO we need code to check the annotations have been set?  All we know at this point is that we have a service with the required label ..... not that it has the annotations as required, with values

Reply:
These 2 annotations must be optional for some extensions that only extends REST API.

Reply:
Another comment here is that if someone comes along and installs an extension after the fact (a new labelled service), the client side UI, presumably, is going to try and register that because your get endpoint would render it

We should persistently store (e.g. in a configmap) the extension at install time *or* use an informer and then users just need to restart their browser to pick it up (until we offer it over a socket).

Reply:
Also bear in mind annotation spec at https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/ doesn't let us spaces, we could client side swap certain chars for spaces if we were desperate...

Reply:
For now (My current issue requirement understanding), the extensions are registered only at the pod start up.  
Thanks for the annotation spec.  I didn't know the restrictions.